core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -247,68 +290,99 @@ ByteBuffer readMaxKey() {
-            return;
-        default:
-            entries[item + offset] = value;
-            return;
+            case KEY_LENGTH:
+                // OFFSET_KEY_LENGTH and OFFSET_KEY_BLOCK should be less then 16 bits long
+                // *2 in order to get read of the signed vs unsigned limits
+                assert value < Short.MAX_VALUE * 2;
+                // set two low bytes of the handle index int
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value & KEY_LENGTH_MASK);
+                return;
+            case KEY_BLOCK:
+                // OFFSET_KEY_LENGTH and OFFSET_KEY_BLOCK should be less then 16 bits long
+                // *2 in order to get read of the signed vs unsigned limits
+                assert value < Short.MAX_VALUE * 2;
+                // offset must be OFFSET_KEY_BLOCK,
+                // set 2 high bytes of the int inside OFFSET_KEY_LENGTH
+                assert value > 0; // block id can never be 0
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value << KEY_BLOCK_SHIFT);
+                return;
+            case VALUE_LENGTH:
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value & VALUE_LENGTH_MASK);
+                return;
+            case VALUE_BLOCK:
+                assert value > 0; // block id can never be 0
+                entries[item + offset.value] =
+                        (entries[item + offset.value]) | (value << VALUE_BLOCK_SHIFT);
+                return;
+            default:
+                entries[item + offset.value] = value;
         }
     }
 
     /**
      * gets the value for the given item, or 'null' if it doesn't exist
      */
-    Handle getHandle(int entryIndex) {
-
-        int hi = getEntryField(entryIndex, OFFSET_HANDLE_INDEX);
+    Slice getValueSlice(int entryIndex) {
 
+        long valueReference = getValueReference(entryIndex);
+        int[] valueArray = UnsafeUtils.longToInts(valueReference);
         // if no value for item - return null
-        assert hi != 0; // because we got to it from linked list
-        if (hi < 0) {
+        if (valueArray[0] == INVALID_BLOCK_ID) {
             return null;
         } else {
-            return handles[hi];
+            return buildValueSlice(valueReference);
+        }
+    }
+
+    Slice buildValueSlice(long valueReference) {
+        int[] valueArray = UnsafeUtils.longToInts(valueReference);
+        if ((valueArray[0] >>> VALUE_BLOCK_SHIFT) == INVALID_BLOCK_ID) {
+            return null;
         }
+        return new Slice(valueArray[0] >>> VALUE_BLOCK_SHIFT, valueArray[1], valueArray[0] & VALUE_LENGTH_MASK,
+                memoryManager);
+    }
+
+    // Assuming the reading of valuePosition and valueBlockAndLength is atomic!
+    long getValueReference(int entryIndex) {
+        int valuePosition, valueBlockAndLength;
```

Comment:
Aren't position+blockId+length updated atomically together with a long CAS?

Reply:
If yes, can't we read them atomically together as long?

Reply:
We can using unsafe. Will be fixed!